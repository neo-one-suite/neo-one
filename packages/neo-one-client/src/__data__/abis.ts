import { ABI, ABIEvent, ABIFunction, ABIParameter, ABIReturn } from '../types';

export const returns = {
  Signature: { type: 'Signature' } as ABIReturn,
  Boolean: { type: 'Boolean' } as ABIReturn,
  Hash160: { type: 'Hash160' } as ABIReturn,
  Hash256: { type: 'Hash256' } as ABIReturn,
  ByteArray: { type: 'ByteArray' } as ABIReturn,
  PublicKey: { type: 'PublicKey' } as ABIReturn,
  String: { type: 'String' } as ABIReturn,
  InteropInterface: { type: 'InteropInterface' } as ABIReturn,
  Void: { type: 'Void' } as ABIReturn,
  Integer: { type: 'Integer', decimals: 7 } as ABIReturn,
  Array: { type: 'Array', value: { type: 'String' } } as ABIReturn,
};

export const parameters = {
  Signature: { name: 'signatureName', type: 'Signature' } as ABIParameter,
  Boolean: { name: 'booleanName', type: 'Boolean' } as ABIParameter,
  Integer: { name: 'intName', type: 'Integer', decimals: 7 } as ABIParameter,
  Hash160: { name: 'hash160Name', type: 'Hash160' } as ABIParameter,
  Hash256: { name: 'hash256Name', type: 'Hash256' } as ABIParameter,
  ByteArray: { name: 'byteName', type: 'ByteArray' } as ABIParameter,
  PublicKey: { name: 'pubkeyName', type: 'PublicKey' } as ABIParameter,
  String: { name: 'stringName', type: 'String' } as ABIParameter,
  Array: {
    name: 'arrayName',
    type: 'Array',
    value: {
      name: 'stringName',
      type: 'String',
    },
  } as ABIParameter,

  InteropInterface: {
    name: 'interopName',
    type: 'InteropInterface',
  } as ABIParameter,
  Void: { name: 'voidName', type: 'Void' } as ABIParameter,
};

export const abiFunction = (
  constant = false,
  params: ABIParameter[] = [parameters.String],
  verify = false,
  name = 'funcName',
  returnType: ABIReturn = returns.Void,
): ABIFunction => ({
  name,
  constant,
  parameters: params,
  verify,
  returnType,
});

export const abiEvent = (name = 'eventName', params: ABIParameter[] = [parameters.String]): ABIEvent => ({
  name,
  parameters: params,
});

export const abi = (functions: ABIFunction[] = [abiFunction()], events: ABIEvent[] = []): ABI => ({
  functions,
  events,
});
