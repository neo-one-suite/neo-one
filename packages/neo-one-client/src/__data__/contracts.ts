import BigNumber from 'bignumber.js';
import { bigNumberToBN } from '../utils';

export const contracts = {
  Signature: {
    type: 'Signature',
    value: '02028a99826ed',
  },
  Boolean: { type: 'Boolean', value: true },
  Integer: {
    type: 'Integer',
    value: bigNumberToBN(new BigNumber('10'), 10),
  },
  Hash160: {
    type: 'Hash160',
    value: '0x3775292229eccdf904f16fff8e83e7cffdc0f0ce',
  },
  Hash256: {
    type: 'Hash256',
    value: '0x798fb9e4f3437e4c9ab83c24f950f0fce98e1d1aaac4fe3f54a66c5d9def89ca',
  },
  ByteArray: {
    type: 'ByteArray',
    value: '02028a99826ed',
  },
  PublicKey: {
    type: 'PublicKey',
    value: '02028a99826edc0c97d18e22b6932373d908d323aa7f92656a77ec26e8861699ef',
  },
  String: {
    type: 'String',
    value: 'stringVal',
  },
  Array: {
    type: 'Array',
    value: [{ type: 'String', value: 'stringVal' }],
  },
  InteropInterface: { type: 'InteropInterface' },
  Void: { type: 'Void' },
};
