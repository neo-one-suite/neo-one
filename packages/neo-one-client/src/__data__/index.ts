export * from './keys';
export * from './transactions';
