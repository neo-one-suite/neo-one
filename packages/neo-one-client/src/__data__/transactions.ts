import { AssetType, common, crypto, RegisterTransaction, ScriptBuilder } from '@neo-one/client-core';

const scriptBuilder = new ScriptBuilder();
scriptBuilder.emitOp('PUSH1');
const admin = crypto.toScriptHash(scriptBuilder.build());
const register = new RegisterTransaction({
  asset: {
    type: AssetType.GoverningToken,
    name: '[{"lang":"zh-CN","name":"小蚁股"},{"lang":"en","name":"AntShare"}]',
    amount: common.fixed8FromDecimal(100000000),
    precision: 0,
    owner: common.ECPOINT_INFINITY,
    admin,
  },
});

export const transactions = {
  register,
};
