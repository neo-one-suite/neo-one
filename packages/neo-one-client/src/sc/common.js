/* @flow */
import { type Param as ScriptBuilderParam } from '@neo-one/client-core';

import _ from 'lodash';

import type {
  Action,
  ActionRaw,
  ABIEvent,
  ABIParameter,
  ABIReturn,
  ContractParameter,
  EventParameters,
  InvocationResult,
  Param,
  RawInvocationResult,
} from '../types';
import {
  InvalidEventError,
  InvalidArgumentError,
  InvocationCallError,
} from '../errors';

import parameterConverters, { converters } from './parameters';
import paramCheckers from './params';

export const convertParameter = ({
  type,
  parameter,
}: {|
  type: ABIParameter | ABIReturn,
  parameter: ContractParameter,
|}): ?Param => parameterConverters[type.type](parameter, (type: $FlowFixMe));

export const getParametersObject = ({
  abiParameters,
  parameters,
}: {|
  abiParameters: Array<ABIParameter>,
  parameters: Array<ContractParameter>,
|}): EventParameters => {
  if (abiParameters.length !== parameters.length) {
    throw new InvalidArgumentError(
      `Expected ABI parameters length (${abiParameters.length}) to equal ` +
        `parameters length (${parameters.length})`,
    );
  }

  return _
    .zip(abiParameters, parameters)
    .reduce((acc, [abiParameter, parameter]) => {
      acc[abiParameter.name] = convertParameter({
        type: abiParameter,
        parameter,
      });
      return acc;
    }, {});
};

export const convertAction = ({
  action,
  events,
}: {|
  action: ActionRaw,
  events: { [key: string]: ABIEvent },
|}): Action => {
  if (action.type === 'Log') {
    return action;
  }

  const { args } = action;
  if (args.length === 0) {
    throw new InvalidEventError('Notification had no arguments');
  }

  const event = converters.toString(args[0]);
  const eventSpec = events[event];
  if (eventSpec == null) {
    throw new InvalidEventError(`Unknown event ${event}`);
  }

  return {
    version: action.version,
    blockIndex: action.blockIndex,
    blockHash: action.blockHash,
    transactionIndex: action.transactionIndex,
    transactionHash: action.transactionHash,
    index: action.index,
    globalIndex: action.globalIndex,
    scriptHash: action.scriptHash,
    type: 'Event',
    name: event,
    parameters: getParametersObject({
      abiParameters: eventSpec.parameters,
      parameters: args.slice(1),
    }),
  };
};

export const convertInvocationResult = ({
  returnType,
  result,
}: {|
  returnType: ABIReturn,
  result: RawInvocationResult,
|}): InvocationResult<?Param> => {
  const { gasConsumed, gasCost } = result;
  if (result.state === 'FAULT') {
    return {
      state: result.state,
      gasConsumed,
      gasCost,
      message: result.message,
    };
  }

  let value = result.stack[0];
  if (value != null) {
    value = convertParameter({
      type: returnType,
      parameter: value,
    });
  }

  return { state: result.state, gasConsumed, gasCost, value };
};

export const convertCallResult = ({
  returnType,
  result,
}: {|
  returnType: ABIReturn,
  result: RawInvocationResult,
|}): ?Param => {
  if (result.state === 'FAULT') {
    throw new InvocationCallError(result.message);
  }

  let value = result.stack[0];
  if (value != null) {
    value = convertParameter({
      type: returnType,
      parameter: value,
    });
  }

  return value;
};

export const convertParams = ({
  parameters,
  params,
}: {|
  parameters: Array<ABIParameter>,
  params: Array<?Param>,
|}): {|
  converted: Array<?ScriptBuilderParam>,
  zipped: Array<[string, ?Param]>,
|} => {
  if (parameters.length !== params.length) {
    throw new InvalidArgumentError(
      `Expected parameters length (${parameters.length}) to equal params ` +
        `length (${params.length}).`,
    );
  }

  const converted = _
    .zip(parameters, params)
    .map(([parameter, param]) =>
      paramCheckers[parameter.type](param, (parameter: $FlowFixMe)),
    );
  const zipped = _
    .zip(parameters, params)
    .map(([parameter, param]) => [parameter.name, param]);

  return { converted, zipped };
};
