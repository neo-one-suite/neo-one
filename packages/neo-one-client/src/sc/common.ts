import { Param as ScriptBuilderParam } from '@neo-one/client-core';
import _ from 'lodash';
import { InvalidArgumentError, InvalidEventError, InvocationCallError } from '../errors';
import {
  ABIEvent,
  ABIParameter,
  ABIReturn,
  Action,
  ActionRaw,
  ContractParameter,
  EventParameters,
  InvocationResult,
  Param,
  RawInvocationResult,
} from '../types';
import { contractParameters, converters } from './parameters';
import { params as paramCheckers } from './params';

export const convertParameter = ({
  type,
  parameter,
}: {
  readonly type: ABIParameter | ABIReturn;
  readonly parameter: ContractParameter;
  // tslint:disable-next-line no-any
}): Param | undefined => (contractParameters[type.type] as any)(parameter, type);

export const getParametersObject = ({
  abiParameters,
  parameters,
}: {
  readonly abiParameters: ReadonlyArray<ABIParameter>;
  readonly parameters: ReadonlyArray<ContractParameter>;
}): EventParameters => {
  if (abiParameters.length !== parameters.length) {
    throw new InvalidArgumentError(
      `Expected ABI parameters length (${abiParameters.length}) to equal ` + `parameters length (${parameters.length})`,
    );
  }

  const zipped = _.zip(abiParameters, parameters) as Array<[ABIParameter, ContractParameter]>;

  return zipped.reduce<EventParameters>(
    (acc, [abiParameter, parameter]) => ({
      ...acc,
      [abiParameter.name]: convertParameter({
        type: abiParameter,
        parameter,
      }),
    }),
    {},
  );
};

export const convertAction = ({
  action,
  events,
}: {
  readonly action: ActionRaw;
  readonly events: { readonly [K in string]?: ABIEvent };
}): Action => {
  if (action.type === 'Log') {
    return action;
  }

  const { args } = action;
  if (args.length === 0) {
    throw new InvalidEventError('Notification had no arguments');
  }

  const event = converters.toString(args[0]);
  const eventSpec = events[event];
  if (eventSpec === undefined) {
    throw new InvalidEventError(`Unknown event ${event}`);
  }

  return {
    version: action.version,
    blockIndex: action.blockIndex,
    blockHash: action.blockHash,
    transactionIndex: action.transactionIndex,
    transactionHash: action.transactionHash,
    index: action.index,
    globalIndex: action.globalIndex,
    scriptHash: action.scriptHash,
    type: 'Event',
    name: event,
    parameters: getParametersObject({
      abiParameters: eventSpec.parameters,
      parameters: args.slice(1),
    }),
  };
};

export const convertInvocationResult = ({
  returnType,
  result,
}: {
  readonly returnType: ABIReturn;
  readonly result: RawInvocationResult;
}): InvocationResult<Param | undefined> => {
  const { gasConsumed, gasCost } = result;
  if (result.state === 'FAULT') {
    return {
      state: result.state,
      gasConsumed,
      gasCost,
      message: result.message,
    };
  }

  const contractParameter = result.stack[0];
  let value;
  // tslint:disable-next-line strict-type-predicates
  if (contractParameter !== undefined) {
    value = convertParameter({
      type: returnType,
      parameter: contractParameter,
    });
  }

  return { state: result.state, gasConsumed, gasCost, value };
};

export const convertCallResult = ({
  returnType,
  result,
}: {
  readonly returnType: ABIReturn;
  readonly result: RawInvocationResult;
}): Param | undefined => {
  if (result.state === 'FAULT') {
    throw new InvocationCallError(result.message);
  }

  const contractParameter = result.stack[0];
  let value;
  // tslint:disable-next-line strict-type-predicates
  if (contractParameter !== undefined) {
    value = convertParameter({
      type: returnType,
      parameter: contractParameter,
    });
  }

  return value;
};

export const convertParams = ({
  parameters,
  params,
}: {
  readonly parameters: ReadonlyArray<ABIParameter>;
  readonly params: ReadonlyArray<Param | undefined>;
}): {
  readonly converted: ReadonlyArray<ScriptBuilderParam | undefined>;
  readonly zipped: ReadonlyArray<[string, Param | undefined]>;
} => {
  if (parameters.length !== params.length) {
    throw new InvalidArgumentError(
      `Expected parameters length (${parameters.length}) to equal params ` + `length (${params.length}).`,
    );
  }

  const zip = _.zip(parameters, params) as Array<[ABIParameter, Param | undefined]>;
  // tslint:disable-next-line no-any
  const converted = zip.map(([parameter, param]) => (paramCheckers[parameter.type] as any)(param, parameter));
  // tslint:disable-next-line no-useless-cast
  const zipped = zip.map(([parameter, param]) => [parameter.name, param] as [string, Param | undefined]);

  return { converted, zipped };
};
