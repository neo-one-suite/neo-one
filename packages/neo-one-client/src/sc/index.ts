export * from './createReadSmartContract';
export * from './createSmartContract';
