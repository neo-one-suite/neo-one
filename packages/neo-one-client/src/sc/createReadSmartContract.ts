import { Param as ScriptBuilderParam } from '@neo-one/client-core';
import { Monitor } from '@neo-one/monitor';
import BigNumber from 'bignumber.js';
import { AsyncIterableX } from 'ix/asynciterable/asynciterablex';
import { filter, map } from 'ix/asynciterable/pipe/index';
import { RawSourceMap } from 'source-map';
import { ReadClient } from '../ReadClient';
import {
  ABIEvent,
  ABIFunction,
  ABIParameter,
  Action,
  ActionRaw,
  BlockFilter,
  Event,
  Hash160String,
  Log,
  Param,
  ReadSmartContract,
  ReadSmartContractDefinition,
  StorageItem,
} from '../types';
import * as common from './common';

const getParams = ({
  parameters,
  params: paramsIn,
}: {
  readonly parameters: ReadonlyArray<ABIParameter>;
  // tslint:disable-next-line no-any
  readonly params: ReadonlyArray<any>;
}): { readonly params: ReadonlyArray<ScriptBuilderParam | undefined>; readonly monitor?: Monitor } => {
  let params = paramsIn;
  const finalArg = params[params.length - 1];
  let monitor;
  if (
    finalArg != undefined &&
    typeof finalArg === 'object' &&
    !Array.isArray(finalArg) &&
    !BigNumber.isBigNumber(finalArg)
  ) {
    params = params.slice(0, -1);
    monitor = finalArg;
  }

  const { converted } = common.convertParams({ params, parameters });

  return { params: converted, monitor };
};

const createCall = ({
  hash,
  client,
  func: { name, parameters = [], returnType },
  sourceMap,
}: {
  readonly hash: Hash160String;
  readonly client: ReadClient;
  readonly func: ABIFunction;
  readonly sourceMap?: RawSourceMap;
  // tslint:disable-next-line no-any
}) => async (...args: any[]): Promise<Param | undefined> => {
  const { params, monitor } = getParams({
    parameters,
    params: args,
  });

  const result = await client.call(hash, name, params, monitor);

  return common.convertCallResult({ returnType, result, sourceMap });
};

export const createReadSmartContract = ({
  definition: {
    hash,
    abi: { events: abiEvents = [], functions },
    sourceMap,
  },
  client,
}: {
  readonly definition: ReadSmartContractDefinition;
  readonly client: ReadClient;
}): ReadSmartContract => {
  const events = abiEvents.reduce<{ [key: string]: ABIEvent }>(
    (acc, event) => ({
      ...acc,
      [event.name]: event,
    }),
    {},
  );

  const iterActionsRaw = (blockFilter: BlockFilter = {}): AsyncIterable<ActionRaw> =>
    // tslint:disable-next-line possible-timing-attack
    AsyncIterableX.from(client.iterActionsRaw(blockFilter)).pipe(filter((action) => action.scriptHash === hash));

  const convertAction = (action: ActionRaw): Action => common.convertAction({ action, events });

  const iterActions = (filterIn?: BlockFilter): AsyncIterable<Action> =>
    AsyncIterableX.from(iterActionsRaw(filterIn)).pipe(map(convertAction));

  const iterEvents = (actionFilter?: BlockFilter): AsyncIterable<Event> =>
    AsyncIterableX.from(iterActions(actionFilter)).pipe(
      map((action) => {
        if (action.type === 'Log') {
          return undefined;
        }

        return action;
      }),
      filter(Boolean),
    );

  const iterLogs = (actionFilter?: BlockFilter): AsyncIterable<Log> =>
    AsyncIterableX.from(iterActions(actionFilter)).pipe(
      map((action) => {
        if (action.type === 'Event') {
          return undefined;
        }

        return action;
      }),
      filter(Boolean),
    );

  const iterStorage = (): AsyncIterable<StorageItem> => client.iterStorage(hash);

  return functions.reduce<ReadSmartContract>(
    (acc, func) =>
      func.constant === true
        ? {
            ...acc,
            [func.name]: createCall({ client, hash, func, sourceMap }),
          }
        : acc,
    {
      iterActionsRaw,
      iterActions,
      iterEvents,
      iterLogs,
      iterStorage,
      convertAction,
    },
  );
};
