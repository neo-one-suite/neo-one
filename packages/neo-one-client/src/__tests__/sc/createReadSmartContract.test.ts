import { toArray } from 'ix/asynciterable/toarray';
import * as abis from '../../__data__/abis';
import * as common from '../../sc/common';
import { createReadSmartContract } from '../../sc/createReadSmartContract';

describe('createReadSmartContract', () => {
  const expectedLog = {
    version: 0,
    blockIndex: 1,
    blockHash: 'blockLogHash',
    transactionIndex: 2,
    transactionHash: 'transLogHash',
    index: 3,
    type: 'Log',
    message: 'logMsg',
    scriptHash: 'hash',
  };

  const expectedEvent = {
    version: 4,
    blockIndex: 5,
    blockHash: 'blockEventHash',
    transactionIndex: 6,
    transactionHash: 'transEventHash',
    index: 7,
    type: 'Event',
    message: 'eventMsg',
    scriptHash: 'hash',
    parameters: {},
    name: 'eventName',
  };

  const expected = [expectedLog, expectedEvent];
  const hash = 'hash';
  const abi = abis.abi([abis.abiFunction(true)], [abis.abiEvent()]);
  const abiNull = {
    functions: [
      {
        name: 'funcName',
        constant: true,
        returnType: abis.returns.Void,
      },

      {
        name: 'undefinedName',
        constant: false,
        returnType: abis.returns.Void,
      },
    ],
  };

  let client = {} as any;

  const verifyMock = (name: string, mock: any) => {
    Object.entries(mock).forEach(([key, maybeMock]) => {
      // @ts-ignore
      if (
        maybeMock != undefined &&
        (maybeMock as any).mock != undefined &&
        (maybeMock as any).mock.calls != undefined
      ) {
        // @ts-ignore
        expect(maybeMock.mock.calls).toMatchSnapshot(`${name}.${key}`);
      }
    });
  };
  const verifyMocks = () => {
    verifyMock('client', client);
    verifyMock('common', common);
  };

  let undefinedEventsContract = createReadSmartContract({
    hash,
    abi: abiNull,
    client,
  });

  let readContract = createReadSmartContract({ hash, abi, client });

  beforeEach(() => {
    client = {} as any;
    // @ts-ignore
    client.call = jest.fn(() => Promise.resolve());
    // @ts-ignore
    (common as any).convertParams = jest.fn(() => expected);
    // @ts-ignore
    (common as any).convertInvocationResult = jest.fn(() => expected);
    // @ts-ignore
    client.iterActionsRaw = jest.fn(() => expected);

    undefinedEventsContract = createReadSmartContract({
      hash,
      abi: abiNull,
      client,
    });

    readContract = createReadSmartContract({ hash, abi, client });
  });

  test('smartContract creation - createCall', async () => {
    // @ts-ignore
    (common as any).convertCallResult = jest.fn(() => expected);

    const result = await readContract[abi.functions[0].name]();
    expect(result).toEqual(expected);

    const undefinedEventsResult = await undefinedEventsContract[abi.functions[0].name]();
    expect(undefinedEventsResult).toEqual(expected);
    expect(undefinedEventsContract.nulName).toBeUndefined();
    verifyMocks();
  });

  test('iterActions with filter', async () => {
    const filter = {
      indexStart: 0,
      indexStop: 1,
    };

    // @ts-ignore
    (common as any).convertAction = jest
      .fn()
      .mockReturnValueOnce(expectedLog)
      .mockReturnValueOnce(expectedEvent);

    const result = await toArray(readContract.iterActions(filter));
    expect(result).toEqual(expected);
    verifyMocks();
  });

  test('iterActions with no filter', async () => {
    // @ts-ignore
    client.iterActionsRaw = jest.fn(() => expected);
    // @ts-ignore
    (common as any).convertAction = jest
      .fn()
      .mockReturnValueOnce(expectedLog)
      .mockReturnValueOnce(expectedEvent);

    const result = await toArray(readContract.iterActions());
    expect(result).toEqual(expected);
    verifyMocks();
  });

  test('iterEvents', async () => {
    // @ts-ignore
    (common as any).convertAction = jest
      .fn()
      .mockReturnValueOnce(expectedLog)
      .mockReturnValueOnce(expectedEvent);

    const result = await toArray(readContract.iterEvents());
    expect(result).toEqual([expectedEvent]);
    verifyMocks();
  });

  test('iterLogs', async () => {
    // @ts-ignore
    (common as any).convertAction = jest
      .fn()
      .mockReturnValueOnce(expectedLog)
      .mockReturnValueOnce(expectedEvent);

    const result = await toArray(readContract.iterLogs());
    expect(result).toEqual([expectedLog]);
    verifyMocks();
  });

  test('iterStorage', async () => {
    // @ts-ignore
    client.iterStorage = jest.fn(() => expected);

    const result = undefinedEventsContract.iterStorage();
    expect(result).toEqual(expected);
    verifyMocks();
  });
});
