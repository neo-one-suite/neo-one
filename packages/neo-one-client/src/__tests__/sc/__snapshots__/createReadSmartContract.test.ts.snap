// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createReadSmartContract iterActions with filter: client.call 1`] = `Array []`;

exports[`createReadSmartContract iterActions with filter: client.iterActionsRaw 1`] = `
Array [
  Array [
    Object {
      "indexStart": 0,
      "indexStop": 1,
    },
  ],
]
`;

exports[`createReadSmartContract iterActions with filter: common.convertAction 1`] = `
Array [
  Array [
    Object {
      "action": Object {
        "blockHash": "blockLogHash",
        "blockIndex": 1,
        "index": 3,
        "message": "logMsg",
        "scriptHash": "hash",
        "transactionHash": "transLogHash",
        "transactionIndex": 2,
        "type": "Log",
        "version": 0,
      },
      "events": Object {
        "eventName": Object {
          "name": "eventName",
          "parameters": Array [
            Object {
              "name": "stringName",
              "type": "String",
            },
          ],
        },
      },
    },
  ],
  Array [
    Object {
      "action": Object {
        "blockHash": "blockEventHash",
        "blockIndex": 5,
        "index": 7,
        "message": "eventMsg",
        "name": "eventName",
        "parameters": Object {},
        "scriptHash": "hash",
        "transactionHash": "transEventHash",
        "transactionIndex": 6,
        "type": "Event",
        "version": 4,
      },
      "events": Object {
        "eventName": Object {
          "name": "eventName",
          "parameters": Array [
            Object {
              "name": "stringName",
              "type": "String",
            },
          ],
        },
      },
    },
  ],
]
`;

exports[`createReadSmartContract iterActions with filter: common.convertCallResult 1`] = `
Array [
  Array [
    Object {
      "result": undefined,
      "returnType": Object {
        "type": "Void",
      },
      "sourceMap": undefined,
    },
  ],
  Array [
    Object {
      "result": undefined,
      "returnType": Object {
        "type": "Void",
      },
      "sourceMap": undefined,
    },
  ],
]
`;

exports[`createReadSmartContract iterActions with filter: common.convertInvocationResult 1`] = `Array []`;

exports[`createReadSmartContract iterActions with filter: common.convertParams 1`] = `Array []`;

exports[`createReadSmartContract iterActions with no filter: client.call 1`] = `Array []`;

exports[`createReadSmartContract iterActions with no filter: client.iterActionsRaw 1`] = `
Array [
  Array [
    Object {},
  ],
]
`;

exports[`createReadSmartContract iterActions with no filter: common.convertAction 1`] = `
Array [
  Array [
    Object {
      "action": Object {
        "blockHash": "blockLogHash",
        "blockIndex": 1,
        "index": 3,
        "message": "logMsg",
        "scriptHash": "hash",
        "transactionHash": "transLogHash",
        "transactionIndex": 2,
        "type": "Log",
        "version": 0,
      },
      "events": Object {
        "eventName": Object {
          "name": "eventName",
          "parameters": Array [
            Object {
              "name": "stringName",
              "type": "String",
            },
          ],
        },
      },
    },
  ],
  Array [
    Object {
      "action": Object {
        "blockHash": "blockEventHash",
        "blockIndex": 5,
        "index": 7,
        "message": "eventMsg",
        "name": "eventName",
        "parameters": Object {},
        "scriptHash": "hash",
        "transactionHash": "transEventHash",
        "transactionIndex": 6,
        "type": "Event",
        "version": 4,
      },
      "events": Object {
        "eventName": Object {
          "name": "eventName",
          "parameters": Array [
            Object {
              "name": "stringName",
              "type": "String",
            },
          ],
        },
      },
    },
  ],
]
`;

exports[`createReadSmartContract iterActions with no filter: common.convertCallResult 1`] = `
Array [
  Array [
    Object {
      "result": undefined,
      "returnType": Object {
        "type": "Void",
      },
      "sourceMap": undefined,
    },
  ],
  Array [
    Object {
      "result": undefined,
      "returnType": Object {
        "type": "Void",
      },
      "sourceMap": undefined,
    },
  ],
]
`;

exports[`createReadSmartContract iterActions with no filter: common.convertInvocationResult 1`] = `Array []`;

exports[`createReadSmartContract iterActions with no filter: common.convertParams 1`] = `Array []`;

exports[`createReadSmartContract iterEvents: client.call 1`] = `Array []`;

exports[`createReadSmartContract iterEvents: client.iterActionsRaw 1`] = `
Array [
  Array [
    Object {},
  ],
]
`;

exports[`createReadSmartContract iterEvents: common.convertAction 1`] = `
Array [
  Array [
    Object {
      "action": Object {
        "blockHash": "blockLogHash",
        "blockIndex": 1,
        "index": 3,
        "message": "logMsg",
        "scriptHash": "hash",
        "transactionHash": "transLogHash",
        "transactionIndex": 2,
        "type": "Log",
        "version": 0,
      },
      "events": Object {
        "eventName": Object {
          "name": "eventName",
          "parameters": Array [
            Object {
              "name": "stringName",
              "type": "String",
            },
          ],
        },
      },
    },
  ],
  Array [
    Object {
      "action": Object {
        "blockHash": "blockEventHash",
        "blockIndex": 5,
        "index": 7,
        "message": "eventMsg",
        "name": "eventName",
        "parameters": Object {},
        "scriptHash": "hash",
        "transactionHash": "transEventHash",
        "transactionIndex": 6,
        "type": "Event",
        "version": 4,
      },
      "events": Object {
        "eventName": Object {
          "name": "eventName",
          "parameters": Array [
            Object {
              "name": "stringName",
              "type": "String",
            },
          ],
        },
      },
    },
  ],
]
`;

exports[`createReadSmartContract iterEvents: common.convertCallResult 1`] = `
Array [
  Array [
    Object {
      "result": undefined,
      "returnType": Object {
        "type": "Void",
      },
      "sourceMap": undefined,
    },
  ],
  Array [
    Object {
      "result": undefined,
      "returnType": Object {
        "type": "Void",
      },
      "sourceMap": undefined,
    },
  ],
]
`;

exports[`createReadSmartContract iterEvents: common.convertInvocationResult 1`] = `Array []`;

exports[`createReadSmartContract iterEvents: common.convertParams 1`] = `Array []`;

exports[`createReadSmartContract iterLogs: client.call 1`] = `Array []`;

exports[`createReadSmartContract iterLogs: client.iterActionsRaw 1`] = `
Array [
  Array [
    Object {},
  ],
]
`;

exports[`createReadSmartContract iterLogs: common.convertAction 1`] = `
Array [
  Array [
    Object {
      "action": Object {
        "blockHash": "blockLogHash",
        "blockIndex": 1,
        "index": 3,
        "message": "logMsg",
        "scriptHash": "hash",
        "transactionHash": "transLogHash",
        "transactionIndex": 2,
        "type": "Log",
        "version": 0,
      },
      "events": Object {
        "eventName": Object {
          "name": "eventName",
          "parameters": Array [
            Object {
              "name": "stringName",
              "type": "String",
            },
          ],
        },
      },
    },
  ],
  Array [
    Object {
      "action": Object {
        "blockHash": "blockEventHash",
        "blockIndex": 5,
        "index": 7,
        "message": "eventMsg",
        "name": "eventName",
        "parameters": Object {},
        "scriptHash": "hash",
        "transactionHash": "transEventHash",
        "transactionIndex": 6,
        "type": "Event",
        "version": 4,
      },
      "events": Object {
        "eventName": Object {
          "name": "eventName",
          "parameters": Array [
            Object {
              "name": "stringName",
              "type": "String",
            },
          ],
        },
      },
    },
  ],
]
`;

exports[`createReadSmartContract iterLogs: common.convertCallResult 1`] = `
Array [
  Array [
    Object {
      "result": undefined,
      "returnType": Object {
        "type": "Void",
      },
      "sourceMap": undefined,
    },
  ],
  Array [
    Object {
      "result": undefined,
      "returnType": Object {
        "type": "Void",
      },
      "sourceMap": undefined,
    },
  ],
]
`;

exports[`createReadSmartContract iterLogs: common.convertInvocationResult 1`] = `Array []`;

exports[`createReadSmartContract iterLogs: common.convertParams 1`] = `Array []`;

exports[`createReadSmartContract iterStorage: client.call 1`] = `Array []`;

exports[`createReadSmartContract iterStorage: client.iterActionsRaw 1`] = `Array []`;

exports[`createReadSmartContract iterStorage: client.iterStorage 1`] = `
Array [
  Array [
    "hash",
  ],
]
`;

exports[`createReadSmartContract iterStorage: common.convertAction 1`] = `
Array [
  Array [
    Object {
      "action": Object {
        "blockHash": "blockLogHash",
        "blockIndex": 1,
        "index": 3,
        "message": "logMsg",
        "scriptHash": "hash",
        "transactionHash": "transLogHash",
        "transactionIndex": 2,
        "type": "Log",
        "version": 0,
      },
      "events": Object {
        "eventName": Object {
          "name": "eventName",
          "parameters": Array [
            Object {
              "name": "stringName",
              "type": "String",
            },
          ],
        },
      },
    },
  ],
  Array [
    Object {
      "action": Object {
        "blockHash": "blockEventHash",
        "blockIndex": 5,
        "index": 7,
        "message": "eventMsg",
        "name": "eventName",
        "parameters": Object {},
        "scriptHash": "hash",
        "transactionHash": "transEventHash",
        "transactionIndex": 6,
        "type": "Event",
        "version": 4,
      },
      "events": Object {
        "eventName": Object {
          "name": "eventName",
          "parameters": Array [
            Object {
              "name": "stringName",
              "type": "String",
            },
          ],
        },
      },
    },
  ],
]
`;

exports[`createReadSmartContract iterStorage: common.convertCallResult 1`] = `
Array [
  Array [
    Object {
      "result": undefined,
      "returnType": Object {
        "type": "Void",
      },
      "sourceMap": undefined,
    },
  ],
  Array [
    Object {
      "result": undefined,
      "returnType": Object {
        "type": "Void",
      },
      "sourceMap": undefined,
    },
  ],
]
`;

exports[`createReadSmartContract iterStorage: common.convertInvocationResult 1`] = `Array []`;

exports[`createReadSmartContract iterStorage: common.convertParams 1`] = `Array []`;

exports[`createReadSmartContract smartContract creation - createCall: client.call 1`] = `
Array [
  Array [
    "hash",
    "funcName",
    undefined,
    undefined,
  ],
  Array [
    "hash",
    "funcName",
    undefined,
    undefined,
  ],
]
`;

exports[`createReadSmartContract smartContract creation - createCall: client.iterActionsRaw 1`] = `Array []`;

exports[`createReadSmartContract smartContract creation - createCall: common.convertCallResult 1`] = `
Array [
  Array [
    Object {
      "result": undefined,
      "returnType": Object {
        "type": "Void",
      },
      "sourceMap": undefined,
    },
  ],
  Array [
    Object {
      "result": undefined,
      "returnType": Object {
        "type": "Void",
      },
      "sourceMap": undefined,
    },
  ],
]
`;

exports[`createReadSmartContract smartContract creation - createCall: common.convertInvocationResult 1`] = `Array []`;

exports[`createReadSmartContract smartContract creation - createCall: common.convertParams 1`] = `
Array [
  Array [
    Object {
      "parameters": Array [
        Object {
          "name": "stringName",
          "type": "String",
        },
      ],
      "params": Array [],
    },
  ],
  Array [
    Object {
      "parameters": Array [],
      "params": Array [],
    },
  ],
]
`;
